src/NuGet.Clients/NuGet.VisualStudio.Common/Services/NuGetFeatureFlagService.cs

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+using Microsoft.Internal.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Shell;
+using NuGet.Common;
+
+namespace NuGet.VisualStudio
+{
+    [Export(typeof(INuGetFeatureFlagService))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class NuGetFeatureFlagService : INuGetFeatureFlagService
+    {
+        private readonly IEnvironmentVariableReader _environmentVariableReader;
+        private readonly IAsyncServiceProvider _asyncServiceProvider;
+        private readonly Microsoft.VisualStudio.Threading.AsyncLazy<IVsFeatureFlags> _ivsFeatureFlags;
+        private readonly Dictionary<string, bool> _featureFlagCache;
+
+        [ImportingConstructor]
+        public NuGetFeatureFlagService()
+            : this(EnvironmentVariableWrapper.Instance, AsyncServiceProvider.GlobalProvider)
+        {
+            // ensure uniqueness.
+        }
+
+        internal NuGetFeatureFlagService(IEnvironmentVariableReader environmentVariableReader, IAsyncServiceProvider asyncServiceProvider)
+        {
+            _environmentVariableReader = environmentVariableReader ?? throw new ArgumentNullException(nameof(environmentVariableReader));
+            _asyncServiceProvider = asyncServiceProvider ?? throw new ArgumentNullException(nameof(asyncServiceProvider));
+            _ivsFeatureFlags = new(() => _asyncServiceProvider.GetServiceAsync<SVsFeatureFlags, IVsFeatureFlags>(), NuGetUIThreadHelper.JoinableTaskFactory);
+            _featureFlagCache = new();
+        }
+
+        public async Task<bool> IsFeatureEnabledAsync(NuGetFeatureFlagConstants experiment)
```

Comment:
It feels like you're getting rid of the concept of an experiment in this PR, should this variable reflect that?  Or is feature flag considered an experiment?  I see both names in this PR.

```suggestion
        public async Task<bool> IsFeatureEnabledAsync(NuGetFeatureFlagConstants featureFlagConstant)
```

Reply:
Yeah, this is supposed to be a preview feature and a lot of the code was duplicated so this was a leftover. I'll fix it.

> Or is feature flag considered an experiment

Regarding this...technically no because you don't get a scorecard and all of that, but it is a roll out strategy.